File: src/units.py
"""Metric/imperial display conversion — direct feedback, 2026-09-13: "Can we
adjust format (metric/imperial) in settings?" (the second of two related remarks,
the first being "can we integrate units into header," which shipped first as a
purely cosmetic change — bare numbers stayed in whatever unit pc caddie/Open-Meteo
already hand back, just labelled in the header. This module is what makes the
*numbers themselves* actually change too.

Pure functions, no I/O, same reasoning `playability.py` already gives for being
implemented ahead of its own real dependency: nothing here needs a live fetch to
exist or be tested.

Deliberately narrow in scope: only the three physical quantities this app ever
displays (temperature, wind speed, precipitation amount) convert. A percentage
(rain probability, occupancy) is already unit-agnostic and never touches this
module. `settings_screen.py`'s own threshold fields (`avoid_temp_below_c`,
`avoid_wind_kph`, `avoid_rain_mm`) are deliberately NOT converted here — those are
stored, compared, and edited in one fixed unit regardless of this display
preference, same as pc caddie's own site always reports occupancy as a plain
fraction regardless of language. Converting stored thresholds bidirectionally
(so an edit in "imperial mode" round-trips correctly) is a real, separate feature,
not implied by "show me the numbers I already see in units I understand."
"""

METRIC = "metric"
IMPERIAL = "imperial"
DEFAULT_UNITS = METRIC

# (label suffix, symbol) shown in a table's own column header -- see tui.py's
# _column_header(). Precipitation pairs a unit-agnostic percentage with an amount,
# hence the "%/mm" vs "%/in" shape rather than a bare unit.
SYMBOLS = {
    METRIC: {"temperature": "°C", "wind": "km/h", "precipitation": "%/mm"},
    IMPERIAL: {"temperature": "°F", "wind": "mph", "precipitation": "%/in"},
}


def celsius_to_fahrenheit(celsius: float) -> float:
    return celsius * 9 / 5 + 32


def kph_to_mph(kph: float) -> float:
    return kph * 0.621371


def mm_to_inches(mm: float) -> float:
    return mm * 0.0393701


def display_temperature(celsius: float, units: str = DEFAULT_UNITS) -> float:
    """`celsius`, converted for display if `units` is imperial -- otherwise
    returned unchanged. Callers still do their own `:.0f`-style rounding at
    format time, same as before this module existed."""
    return celsius_to_fahrenheit(celsius) if units == IMPERIAL else celsius


def display_wind_speed(kph: float, units: str = DEFAULT_UNITS) -> float:
    return kph_to_mph(kph) if units == IMPERIAL else kph


def display_precipitation_mm(mm: float, units: str = DEFAULT_UNITS) -> float:
    return mm_to_inches(mm) if units == IMPERIAL else mm


def precipitation_amount_label(units: str = DEFAULT_UNITS) -> str:
    """The only remaining per-cell unit label (2026-09-13, direct follow-up:
    "since the units are now in the headers, we don't need the units in the
    rows, right?") -- temperature/wind cells dropped their own "°"/"km/h"
    suffixes once the column header started stating the unit (a
    `wind_unit_label()` counterpart used to live here too), but a
    precipitation cell packs two different numbers together ("70%/1.5mm"), so
    this one still earns its keep: it's what tells the second number apart
    from the first, not just a repeat of the header."""
    return "in" if units == IMPERIAL else "mm"

File: prototypes/macos-swift/Sources/Units.swift
import Foundation

/// Metric/imperial display conversion -- ports `src/units.py` exactly, including
/// its deliberate narrowness: only the physical quantities this app actually
/// displays convert, and a percentage (rain probability, occupancy) is already
/// unit-agnostic and never touches this.
///
/// **Why this exists**: Settings has had a Units picker since Tier 1, and it wrote
/// `units:` to `preferences.yaml` correctly the whole time -- but nothing in this
/// app ever *read* it back. Every temperature was a hardcoded `"%.0f°"` and every
/// wind speed a bare km/h number, so switching to imperial changed the file, the
/// TUI, and nothing whatsoever on screen here. Reported directly: "changing metric
/// to imperial also doesn't change anything." The setting was never broken; the
/// display simply ignored it.
enum Units {
    static let metric = "metric"
    static let imperial = "imperial"

    static func celsiusToFahrenheit(_ c: Double) -> Double { c * 9 / 5 + 32 }
    static func kphToMph(_ kph: Double) -> Double { kph * 0.621371 }
    static func mmToInches(_ mm: Double) -> Double { mm * 0.0393701 }

    static func temperature(_ celsius: Double, _ units: String) -> Double {
        units == imperial ? celsiusToFahrenheit(celsius) : celsius
    }

    static func windSpeed(_ kph: Double, _ units: String) -> Double {
        units == imperial ? kphToMph(kph) : kph
    }

    static func precipitationMM(_ mm: Double, _ units: String) -> Double {
        units == imperial ? mmToInches(mm) : mm
    }

    /// Mirrors `units.SYMBOLS` -- used by the legend line, which is this app's
    /// equivalent of the TUI's own unit-bearing column headers.
    static func temperatureSymbol(_ units: String) -> String { units == imperial ? "°F" : "°C" }
    static func windSymbol(_ units: String) -> String { units == imperial ? "mph" : "km/h" }

    /// Mirrors `units.precipitation_amount_label()` -- the one per-cell unit
    /// suffix that survived the header taking over every other unit label (see
    /// that function's own docstring): a precipitation cell packs two numbers
    /// together ("70%/1.5mm"), so this is what tells them apart. Added
    /// 2026-09-19 alongside actually displaying the mm/in amount at all (direct
    /// report, "precipitation amount in mm seems to still be missing") --
    /// SlotRow/SearchResultRow previously showed only the probability
    /// percentage, never the amount.
    static func precipitationAmountLabel(_ units: String) -> String { units == imperial ? "in" : "mm" }
}

/// The running app's current units -- same singleton shape as `AppTheme`/`AppScale`,
/// for the same reason: `SettingsSheet` assigning to it is what redraws every view
/// showing a temperature or wind speed, immediately and with no relaunch.
///
/// Seeded from `preferences.yaml` (the same file the TUI reads), not from its own
/// key, so the two front ends genuinely share one setting rather than each keeping
/// a private copy.
final class AppUnits: ObservableObject {
    static let shared = AppUnits()

    @Published var value: String

    private init() {
        value = Preferences.load().units
    }
}
